src/main/host/thread.rs

```diff
@@ -0,0 +1,72 @@
+use super::syscall_types::{PluginPtr, SysCallReg};
+use crate::cbindings as c;
+use crate::utility::syscall;
+use libc;
+
+/// Wraps the C Thread struct.
+pub struct Thread {
+    cthread: *mut c::Thread,
+}
+
+impl Thread {
+    pub fn new(cthread: *mut c::Thread) -> Thread {
+        unsafe {
+            c::thread_ref(cthread);
+        }
+        Thread { cthread }
+    }
+
+    fn native_syscall(&mut self, n: i64, args: &[SysCallReg]) -> Result<SysCallReg, i32> {
+        let raw_res;
+        let mut i = args.iter().copied();
+        // The `Into` here isn't strictly necessary since SysCallReg is an alias for c::SysCallReg,
+        // but that's an implementation detail of SysCallReg that could change later, and the
+        // compiler won't help us catch it when calling a variadic C function.
+        let mut a = || Into::<c::SysCallReg>::into(i.next().unwrap());
+        unsafe {
+            raw_res = match args.len() {
+                0 => c::thread_nativeSyscall(self.cthread, n),
+                1 => c::thread_nativeSyscall(self.cthread, n, a()),
+                2 => c::thread_nativeSyscall(self.cthread, n, a(), a()),
+                3 => c::thread_nativeSyscall(self.cthread, n, a(), a(), a()),
+                4 => c::thread_nativeSyscall(self.cthread, n, a(), a(), a(), a()),
+                5 => c::thread_nativeSyscall(self.cthread, n, a(), a(), a(), a(), a()),
+                6 => c::thread_nativeSyscall(self.cthread, n, a(), a(), a(), a(), a(), a()),
```

Comment:
Oh wow, that didn't even occur to me, and I've been bitten by that before in C where it can and does vary by compiler and flags. Rust is making me too complacent ;)

I poked around the link you provided and other discussions and agree this *probably* is safe, but decided to just go back to explicit indexing to avoid this issue.